### PR TITLE
DevTools: Support mulitple DevTools instances per page

### PR DESCRIPTION
**Pull request by bvaughn**
This is being done so that we can embed DevTools within the new React (beta) docs.

The primary changes here are to `react-devtools-inline/backend`:
* Add a new `createBridge` API
* Add an option to the `activate` method to support passing in the custom bridge object.

To verify these changes, this commit also updates the test shell to add a new entry-point for multiple DevTools.

This PR should unblock https://github.com/reactjs/reactjs.org/pull/4161/

### CODE REVIEW

File: packages/react-devtools-inline/src/backend.js [hunk 1]
```diff
@@ -61,27 +64,11 @@ function startActivation(contentWindow: window) {
   // because they are stored in localStorage within the context of the extension (on the frontend).
   // Instead it relies on the extension to pass preferences through.
   // Because we might be in a sandboxed iframe, we have to ask for them by way of postMessage().
+  // TODO WHAT HUH
```
Comment (bvaughn) [inline packages/react-devtools-inline/src/backend.js:67]: Whoops. Left this note to myself to follow up on this and then forgot to. Will do momentarily.

File: packages/react-devtools-shared/src/backend/agent.js [hunk 2]
```diff
@@ -320,6 +323,7 @@ export default class Agent extends EventEmitter<{|
   }
 
   getBridgeProtocol = () => {
+    console.log('[agent] getBridgeProtocol -> bridge.send("bridgeProtocol")');
```
Comment (lunaruan) [inline packages/react-devtools-shared/src/backend/agent.js:326]: Intentional?

File: packages/react-devtools-inline/README.md [hunk 3]
```diff
@@ -177,32 +177,47 @@ Below is an example of an advanced integration with a website like [Replay.io](h
...
+  createBridge as createFrontendBridge,
   createStore,
   initialize as createDevTools,
-} from "react-devtools-inline/frontend";
+} from 'react-devtools-inline/frontend';
 
-// Custom Wall implementation enables serializing data
-// using an API other than window.postMessage()
+// DevTools uses "message" events and window.postMessage() by default,
+// but we can override this behavior by creating a custom "Wall" object.
```
Comment (Manndung75) [inline packages/react-devtools-inline/README.md:191]: A

File: packages/react-devtools-shared/src/backend/agent.js (hunk 2 above)
Comment (bvaughn) [reply packages/react-devtools-shared/src/backend/agent.js:326]: Oops, no

**Comment by bvaughn** [comment]
I had to push an update that removed two direct calls to `window.postMessage` and replace them with `bridge.send`. Should have done this a long time ago but just hadn't noticed until I tried replacing the `postMessage` API with a custom bridge.

This makes our Flow type coverage a little better for the get/set saved preferences stuff too that frontend and backend exchange during startup.

I'll do a little more testing to make sure there's nothing else like this.

File: packages/react-devtools-inline/src/frontend.js [hunk 4]
```diff
@@ -68,49 +64,40 @@ export function initialize(
...
+
   if (store == null) {
-    store = createStore(bridge);
+    store = createStore(frontendBridge);
   }
 
+  const onGetSavedPreferences = () => {
+    // This is the only message we're listening for,
+    // so it's safe to cleanup after we've received it.
+    frontendBridge.removeListener('getSavedPreferences', onGetSavedPreferences);
```
Comment (bvaughn) [inline packages/react-devtools-inline/src/frontend.js:81]: This should have been using the bridge all along. (Was an oversight.)

File: packages/react-devtools-shared/src/backend/agent.js [hunk 5]
```diff
@@ -225,6 +225,9 @@ export default class Agent extends EventEmitter<{|
       bridge.send('profilingStatus', true);
     }
 
+    // Send the Bridge protocol after initialization in case the frontend has already requested it.
+    this._bridge.send('bridgeProtocol', currentBridgeProtocol);
```
Comment (bvaughn) [inline packages/react-devtools-shared/src/backend/agent.js:229]: Made this change to avoid potential timing problems between the backend and frontend initialization sequence. (If the frontend asked the backend for a protocol _before_ the backend was listening, it would cause the frontend to timeout waiting and show a modal:
Screen Shot 2021-12-14 at 9 17 31 AM

File: packages/react-devtools-inline/README.md [hunk 6]
```diff
@@ -177,32 +177,47 @@ Below is an example of an advanced integration with a website like [Replay.io](h
...
   },
 };
 
-// Create a Bridge and Store that use the custom Wall.
+// Initialize the DevTools backend before importing React (or any other packages that might import React).
+initializeBackend(contentWindow);
+
+// Prepare DevTools for rendering.
+// To use the custom Wall we've created, we need to also create our own "Bridge" and "Store" objects.
 const bridge = createBridge(target, wall);
```
Comment (jstejada) [inline packages/react-devtools-inline/README.md:208]: should this be `createFrontendBridge` now?

File: packages/react-devtools-shell/now.json [hunk 7]
```diff
@@ -1,5 +0,0 @@
-{
-  "name": "react-devtools-experimental",
```
Comment (jstejada) [inline packages/react-devtools-shell/now.json:2]: what was this file for?

File: packages/react-devtools-shared/src/backend/agent.js (hunk 5 above)
Comment (jstejada) [reply packages/react-devtools-shared/src/backend/agent.js:229]: does this change affect the extension or standalone apps in any way? should we test that?

File: packages/react-devtools-inline/src/backend.js [hunk 8]
```diff
@@ -5,83 +5,57 @@ import Bridge from 'react-devtools-shared/src/bridge';
...
-      default:
-        break;
+import type {BackendBridge} from 'react-devtools-shared/src/bridge';
+import type {Wall} from 'react-devtools-shared/src/types';
+
+function startActivation(contentWindow: window, bridge: BackendBridge) {
+  const onSavedPreferences = data => {
+    // This is the only message we're listening for,
+    // so it's safe to cleanup after we've received it.
+    bridge.removeListener('savedPreferences', onSavedPreferences);
```
[the remark this answers is not in the document]
Comment (jstejada) [reply packages/react-devtools-inline/src/backend.js:16]: instead of postmessage you mean?

File: packages/react-devtools-inline/src/backend.js [hunk 9]
```diff
@@ -5,67 +5,54 @@ import Bridge from 'react-devtools-shared/src/bridge';
...
   // The backend may be unable to read saved preferences directly,
   // because they are stored in localStorage within the context of the extension (on the frontend).
   // Instead it relies on the extension to pass preferences through.
   // Because we might be in a sandboxed iframe, we have to ask for them by way of postMessage().
-  // TODO WHAT HUH
-  parent.postMessage({type: MESSAGE_TYPE_GET_SAVED_PREFERENCES}, '*');
+  bridge.send('getSavedPreferences');
 }
 
 function finishActivation(contentWindow: window, bridge: BackendBridge) {
```
Comment (jstejada) [inline packages/react-devtools-inline/src/backend.js:58]: is `: window` a type here?

File: packages/react-devtools-inline/src/frontend.js [hunk 10]
```diff
@@ -68,49 +64,40 @@ export function initialize(
...
+      hideConsoleLogsInStrictMode: getHideConsoleLogsInStrictMode(),
+    };
+
+    // The renderer interface can't read saved preferences directly,
+    // because they are stored in localStorage within the context of the extension.
+    // Instead it relies on the extension to pass them through.
+    frontendBridge.send('savedPreferences', data);
+  };
+
+  frontendBridge.addListener('getSavedPreferences', onGetSavedPreferences);
```
Comment (jstejada) [inline packages/react-devtools-inline/src/frontend.js:97]: don't really follow everything that's happening with these changes, but i trust you. is this related to the original change in the PR? if not it could've been nice to separate this, but not a big deal

**Review by jstejada — APPROVED** [review]
looks good to me! just left a few questions.. was also curious, maybe you could add a screenshot of what the multi devtools looks like in the test plan?

File: packages/react-devtools-inline/src/backend.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/react-devtools-inline/src/backend.js:16]: Yes. Instead of "message" events

File: packages/react-devtools-shared/src/backend/agent.js (hunk 5 above)
Comment (bvaughn) [reply packages/react-devtools-shared/src/backend/agent.js:229]: I tested the extension. This will only affect RN though.

File: packages/react-devtools-inline/src/backend.js (hunk 9 above)
Comment (bvaughn) [reply packages/react-devtools-inline/src/backend.js:58]: It's the Flow type.

File: packages/react-devtools-inline/src/frontend.js (hunk 10 above)
Comment (bvaughn) [reply packages/react-devtools-inline/src/frontend.js:97]: Yes, it's required to support a custom wall that doens't use "message" events, which is required to support multiple DevTools per page.

File: packages/react-devtools-shell/now.json (hunk 7 above)
Comment (bvaughn) [reply packages/react-devtools-shell/now.json:2]: I used to deploy this demo to Now (Vercel) but we don't anymore.

File: packages/react-devtools-inline/README.md (hunk 6 above)
Comment (bvaughn) [reply packages/react-devtools-inline/README.md:208]: No. It's still exported as `createBridge`

**Comment by bvaughn** [comment]
Thanks for the reviews, @jstejada and @lunaruan !

Here's a screenshot of the multi DevTools harness, Juan:
Screen Shot 2021-12-14 at 12 15 44 PM
